New offboard estimator interface

mhkabir opened the pull request:
This PR brings in the new offboard estimator interface for VIO applications. 

![rectangularconcernedaustraliancattledog](https://cloud.githubusercontent.com/assets/5557676/21762602/989625e0-d67f-11e6-9e99-095c0f95d644.gif)

- The `LOCAL_POSITION_NED_COV` and `ATTITUDE_QUATERNION_COV` messages are used to transfer the **full state estimate** from an offboard EKF (e.g a VIO framework). This includes position,  velocity, acceleration, attitude, angular velocity and covariances.
- The `vision_position_estimate` topic is deprecated in favour of the new `vehicle_vision_position` and `vehicle_vision_attitude` topics (same struct as `local_position` and `vehicle_attitude`). The controllers can use it directly if desired in the future. Works with a cascaded setup too, with an onboard estimator like EKF2/LPE fusing the vision estimate.
- Full backwards compatibility with the `VISION_POSITION_ESTIMATE` message is maintained.
- `VISION_POSITION_NED` debug stream from PX4 is renamed to `VISION_POSITION_ESTIMATE` in line with the mavlink naming. This was previously confusing.
- Estimators updated to subscribe to new topic.
- The `GPS_GLOBAL_ORIGIN` message can now be sent to PX4 to set the local frame global reference in case a offboard estimator fuses GPS. The origin is published in the `vehicle_vision_position` topic, so that an onboard estimator in the loop can initialise its origin to the same reference.
- Currently, nothing takes advantage of the covariances. The plan is to sync with @jgoppert's new IEKF such that all the data is used effectively.
- Needs new logger to be used to log all useful data. We should also add plots to FlightReview with this information. @bkueng ping.

It will be great if you can crosstest and quickly get this merged, because it is primarily re-plumbing changes with no functional differences yet. The diff might also get hard to maintain if left to rot.

Please review! @LorenzMeier @ChristophTobler

mhkabir commented:
@LorenzMeier I don't seem to have permissions to request reviews and assign people. Can you have a look?

mhkabir commented:
CI fail due to old attitude_estimator_ekf. I mistakenly thought it was no longer built. Do we even want to maintain that?

LorenzMeier commented:
It would not be in the CI system if we didn't want to maintain at least its build ability.

LorenzMeier commented:
There are also other structure things I need to look at, but will need to do that with more time. Thanks for the PR!

mhkabir commented:
@LorenzMeier CI fixed. Please let me know if any structural changes are required, because I would like to get this tested and in during this holiday season.

mhkabir commented:
@ecmnet Can you give this a test run with your vision setup. No changes needed to your vision/sender code anywhere. I just want to confirm that there are no regressions in the backward-compatibility code (for `VISION_POSITION_ESTIMATE`) here. 

Vision should work like normal master does, but do the routing slightly differently internally.

Thanks! A quick review and any comments would be welcome too!

ecmnet commented:
@mhkabir With current master, the companion link is broken for me. I will check that as soon as I get the companion link up and running again

LorenzMeier commented:
This looks quite complete and nice. Could you provide some experimental results / testing plots?

The code looks good from a review perspective, I just want to understand what the testing level is and see how it performs in the closed loop.

mhkabir commented:
Definitely, I will try to get some flying in (regulations making it hard ATM), or get you some plots from a hand-held stereo-inertial sensor with a Pixhawk on it. It was tested in a prior state with over 5 hours of VIO-only and VIO-GPS flight, before I cleaned it up for upstreaming. 

I would also like to upstream our changes which were needed to passthrough raw GPS data (with full covariance) to the offboard estimator, and add the hooks in PX4 to choose between full-offboard estimation (VIO passed straight to controllers) or the onboard-estimator-in-loop setup. But that can be a different PR, unless you would like it in here.

mhkabir commented:
@ecmnet If you got your companion link working, it would be great to validate that the legacy interface still performs well. Thanks!

ecmnet commented:
@mhkabir Would be great, if you could post some logs. As I am off now for several days, I am not able to test it asap, but I will do this after I returned.

mhkabir commented:
@LorenzMeier @jgoppert @ecmnet So I made the synchronized stereo-inertial sensor rig I was talking about. Also has GPS! :)

![photo_2017-01-02_20-23-28](https://cloud.githubusercontent.com/assets/5557676/21591382/5ab59be4-d129-11e6-95c3-69c0f4d1d9e8.jpg)

Still need to do some calibrations, and then will be able to provide ground-truth + raw VIO estimates + Onboard LPE estimate plots.

jgoppert commented:
Looks awesome!

On Jan 2, 2017 9:55 AM, "Kabir Mohammed" <notifications@github.com> wrote:

> @LorenzMeier <https://github.com/LorenzMeier> @jgoppert
> <https://github.com/jgoppert> @ecmnet <https://github.com/ecmnet> So I
> made the synchronized stereo-inertial sensor rig I was talking about. Also
> has GPS! :)
>
> [image: photo_2017-01-02_20-23-28]
> <https://cloud.githubusercontent.com/assets/5557676/21591382/5ab59be4-d129-11e6-95c3-69c0f4d1d9e8.jpg>
>
> Still need to do some calibrations, and then will be able to provide
> ground-truth + raw VIO estimates + Onboard LPE estimate plots.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PX4/Firmware/pull/6074#issuecomment-269982511>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAc6rG1lDIH5FW_Ed0a9ptg7h10N800Tks5rOQ_-gaJpZM4LPZuo>
> .
>

mhkabir commented:
Lots of progress made in making this path easier for everyone else. 

1. Proper IMU noise parameters

@jgoppert and I made nice progress in modelling our IMU noise parameters using Allan Deviation plots. The full iPython notebook and results are available here: https://github.com/ProjectArtemis/SensorAnalysis/blob/master/Allan_Deviation.ipynb

![download](https://cloud.githubusercontent.com/assets/5557676/21750290/ddbfb47e-d5d5-11e6-81e4-6fd6e2698331.png)

![download 1](https://cloud.githubusercontent.com/assets/5557676/21750291/e95ff37a-d5d5-11e6-9330-71a37612f1af.png)

These are super-useful for use in spatio-temporal calibration settings where you need to calibrate the extrinsics of IMU-camera rig, e.g using Kalibr (https://github.com/ethz-asl/kalibr). I will also make a Devguide page with information on making and setting up a VI-rig from scratch using a Pixhawk.

2. Fixed long-term Mavros frame rotation bug (Woohoo!)

There was a missing body to inertial transform in the vision plugin. I have fixed it in our custom estimator_bridge plugin, and will eventually rejig the vision plugin to support this new API along with the fix. However, I still need to derive the covariance matrix transformations to make it complete. I am currently out of time becuase of "life" issues, but will return to this soon.

3. Plots with VIO running online - coming tomorrow.

mhkabir commented:
@LorenzMeier @jgoppert @ecmnet Results are here, and they're amazing!

![rectangularconcernedaustraliancattledog](https://cloud.githubusercontent.com/assets/5557676/21762602/989625e0-d67f-11e6-9e99-095c0f95d644.gif)

Ground-truth is calculated using AprilSLAM with IMU preintegration (https://github.com/ProjectArtemis/aprilslam) and a single AprilTag in the scene. (Offset between tag frame and VIO frame is calibrated online.) MultiVIO is running in monocular mode, with the other camera in the stereo rig disabled.

# LPE Position vs. Groundtruth

![screenshot from 2017-01-09 14 59 12](https://cloud.githubusercontent.com/assets/5557676/21762133/0ff6ace8-d67d-11e6-9295-f905dc40b538.png)

# LPE Velocity

![screenshot from 2017-01-09 15 11 08](https://cloud.githubusercontent.com/assets/5557676/21762289/ec68aa1e-d67d-11e6-92d3-b3973ea04e5f.png)

This goes to show that one can go a very long way without special/expensive hardware. This was on a system which was hand-built for $300 or so.

maciejmatuszak commented:
I am putting my quad together after last flight in the mine, lots of dust everywhere... I think I figure-out the mavros rotations to put VIO to PX4 but after rebuild I had to re-tune it again the VIO lost its accuracy/reliability. I will test it as soon as my VIO is back operational.
@mhkabir @jgoppert I was trying to use the Allan Deviation plots but the latest version is not released yet. I was trying to us Anaconda tools to build latest but can not resolve the dependencies, the build complains there is no linux x64 version available for transforms3d and and one more module. Tried on Windows but with the same results.
Are you using Anaconda or different environment?

maciejmatuszak commented:
Maciej, you can install the any version of px4tools in editable mode using:
`pip install -e .`
from within px4tools directory.

That sorted, @mhkabir I assume the sample data for the plot is from stationary IMU as per comment in the px4tools scripts. Will the acceleration noise from running motors and props affect the outcome? Should the sample be taken with motors running?

mhkabir commented:
You will need to multiply them by a factor of 10-100 or so from my experience to capture all the errors. 
And yes, vibrations will affect estimation accuracy. Ideally, your isolation solution must give you under +-1G accelerations in flight for good performance. Might need empirically tune further in case there are more vibrations.

maciejmatuszak commented:
I really do not like python environment, made mess with mixing version 2 and 3 libraries... resulted in rebuilding linux machine..
I am attempting to take sufficient sample for Allan plots, and notice that while the onboard computer (odroid) fan is working the acceleration noise is increasing. Have a look at the Y axes acceleration.
![screenshot from 2017-01-14 19 04 05](https://cloud.githubusercontent.com/assets/2688875/21953523/ef39e7f8-da8d-11e6-926c-e4d873b45386.png)
 Is it a problem? My understanding is that the averaging times are much longer than the 10-20 seconds it should not affect the outcome?
It shows only on Y axes, X and Z seems to be clean, I think it is the way I positioned the anti-vibration foam.
What's the minimum length of a sample? I saw 8 hours in some other discussions?

mhkabir commented:
@maciejmatuszak Can we have this discussion someplace better? It doesn't directly have a bearing on the PR, and I prefer to keep the discussion on-topic. :)


> I really do not like python environment, made mess with mixing version 2 and 3 libraries... resulted in rebuilding linux machine.

Use anaconda3 like recommended in the px4tools readme, and source the environment only when needed. It will mess with ROS otherwise.

maciejmatuszak commented:
@mhkabir Thsnks for answer! Yes I got carried away.

mhkabir commented:
@LorenzMeier I've rebased this. Anything holding this one back? 

I will make subsequent PRs with : 
1. GPS interface (waiting on https://github.com/PX4/Firmware/issues/6275)
2. Copying of covariance matrix and using in EKF2/LPE update step. (see https://github.com/PX4/Firmware/pull/6074#issuecomment-271151886)
3. Modern groundtruth message. (mavlink/mavlink#660)

jgoppert commented on cmake/configs/nuttx_px4fmu-v2_default.cmake:
```diff
@@ -51,7 +51,7 @@ set(config_module_list
 	drivers/lis3mdl
 	#drivers/iridiumsbd
-	drivers/ulanding
+	#drivers/ulanding
```
Not sure what this is, will it be missed?

dagar replied:
Aerotenna uLanding radar added by Roman recently. It should fit now though.

jgoppert commented on msg/vehicle_attitude.msg:
```diff
@@ -4,4 +4,4 @@ float32 pitchspeed	# Angular velocity about body east axis (y) in rad/s
 float32[4] q		# Quaternion (NED)
 
-# TOPICS vehicle_attitude vehicle_attitude_groundtruth
+# TOPICS vehicle_attitude vehicle_attitude_groundtruth vehicle_vision_attitude
```
How is this going to extend for mocap? We run out of topics at 3 right?

mhkabir replied:
Mocap will be going into `vehicle_attitude(/position)_groundtruth`, so its not an issue.

jgoppert commented on msg/vehicle_local_position.msg:
```diff
@@ -47,4 +60,4 @@ bool dist_bottom_valid			# true if distance to bottom surface is valid
 float32 epv				# Standard deviation of vertical position error, (metres)
 
-# TOPICS vehicle_local_position vehicle_local_position_groundtruth
+# TOPICS vehicle_local_position vehicle_local_position_groundtruth vehicle_vision_position
```
same comment above about mocap etc.

jgoppert commented on msg/vehicle_local_position.msg:
```diff
@@ -28,6 +36,11 @@ uint8 z_reset_counter
+# Acceleration in NED frame
+float32 ax        # North acceleration in NED earth-fixed frame, (metres/sec^2)
+float32 ay        # East acceleration in NED earth-fixed frame, (metres/sec^2)
+float32 az        # Down acceleration in NED earth-fixed frame, (metres/sec^2)
```
I believe this is in control_state, do we need it here?

mhkabir replied:
Yes. We want it here because the mavlink message LOCAL_POSITION_NED_COV also has it, and an external estimator (especially VIO) is in the best place to estimate accelerations.

jgoppert commented on src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp:
```diff
@@ -375,10 +376,14 @@ void BlockLocalPositionEstimator::update()
 
 	// if we have no lat, lon initialize projection at 0,0
-	if ((_estimatorInitialized & EST_XY) && !_map_ref.init_done) {
+	if (!_map_ref.init_done && (_estimatorInitialized & EST_XY) && _fake_origin.get()) {
```
My fault, but this comment should be changed to say LPE_LAT, LPE_LON param.

mhkabir replied:
Fixed.

jgoppert commented on src/modules/local_position_estimator/sensors/vision.cpp:
```diff
@@ -72,14 +79,33 @@ void BlockLocalPositionEstimator::visionCorrect()
 	uint8_t i_hist = 0;
 
-	if (getDelayPeriods(_vision_delay.get(), &i_hist)  < 0) { return; }
+	float vision_delay = (_timeStamp - _sub_vision_pos.get().timestamp) * 0.000001f; // measurement delay in seconds
```
would be a bit easier to read as * 1e-6f

mhkabir replied:
Fixed

mhkabir commented:
All comments addressed, rebased on master and force-pushed.

jgoppert commented:
I'd like to flight test before merging with manual/ gps/ flow to make sure we don't have an regressions.

LorenzMeier commented:
@jgoppert @mhkabir Let me know when you flew this (and please share flight logs).

mzahana commented:
Hi. This is very interesting PR. Is it flight tested already?
 Is there a guide on what is required in order test this PR?
Does this have some interface to ROS, e.g. via MAVROS?

LorenzMeier commented:
Rebased and applied. Awesome work Kabir!

mzahana commented:
Great!

@mhkabir can you please provide a guide of how it works?

mhkabir commented:
Will do! I will update DevGuide with information on the new interface. Meanwhile, you can continue to use the old `VISION_POSITION_ESTIMATE` one.

priseborough commented:
@mhkabir This should not have been merged. It breaks the ekf2 replay that also relies on this topic.

mhkabir commented:
I'm sorry, I didn't realise that. How did this PR break it?

priseborough commented:
ekf2_replay_main.cpp uses that topic and won't compile. Check with Beat before fixing it - I think he may already have something. It is good practice to do a  text string search in the file system for the all subscribers/publishers to a topic when making changes to the uORB.

bkueng commented:
@mhkabir I'll push my fix soon, then you can review it.

mhkabir commented:
I expected that it would be in CI, ergo I didn't catch it. Many thanks @bkueng !

mzahana commented:
Any tutorials on how to use this with monocular camera connected to a companion computer e.g. odroid xu4 running Mavros? Are there any other equipment needed?
